contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,816 @@
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`RollingUpdateStatefulSetStrategyType`, then an update is complete when the 
+StatefulSet is at its [target state](#target-state). The StatefulSet controller 
+will signal update completion as follows.
+    1. The controller will set `.Status.CurrentVersion` to the value of 
+    `.Staus.UpdateVersion`.
+    1. The controller will set `.Status.CurrentReplicas` to 
+    `.Status.UpdateReplicas`. Note that this value will be equal to 
+    `.Status.Replicas`.
+    1. The controller will set `.Status.UpdateReplicas` to 0.
+
+### Status Reporting
+After processing the creation, update, or deletion of a StatefulSet or Pod, 
+the StatefulSet controller will record its status by persisting a 
+StatefulSetStatus object. This has two purposes.
+
+1. It allows the StatefulSet controller to recreate the exact StatefulSet 
+membership in the event of a hard restart of the entire system.
+1. It communicates the current state of the StatefulSet to clients. Using the 
+`.Status.ObserverGeneration`, clients can construct a linearizable view of 
+the operations performed by the controller.
+
+When the StatefulSet controller records the status of a StatefulSet it will 
+do the following.
+
+1. The controller will increment the `.Status.ObservedGeneration` to communicate 
+the `.Generation` of the StatefulSet object that was observed.
+1. The controller will set the `.Status.Replicas` to the current number of 
+created Pods.
+1. The controller will set the `.Status.ReadyReplicas` to the current number of 
+Pods that have a Ready Condition.
+1. The controller will set the `.Status.CurrentVersion` and 
+`.Status.UpdateVersion` in accordance with StatefulSet's 
+[revision history](#statefulset-revision history) and 
+any [complete updates](#update-completion).
+1. The controller will set the `.Status.CurrentReplicas` to the number of 
+Pods that it has created from the version indicated by 
+`.Status.CurrentVersion`.
+1. The controller will set the `.Status.UpdatedReplicas` to the number of Pods 
+that it has created from the version indicated by `.Status.UpdateVersion`.
+1. The controller will then persist the StatefulSetStatus make it durable and 
+communicate it to observers.
+
+## API Server
+The API Server will perform validation for StatefulSet creation and updates.
+
+### StatefulSet Validation
+As is currently implemented, the API Server will not allow mutation to any 
+fields of the StatefulSet object other than `.Spec.Replicas` and 
+`.Spec.Template.Containers`. This design imposes the following, additional 
+constraints.
+
+1. If the `.Spec.UpdateStrategy.Type` is equal to 
+`RollingUpdateStatefulSetStrategyType`, the API Server should fail validation 
+if any of the following conditions are true.
+   1. `.Spec.UpdateStrategy.Partition` is nil.
+   1. `.Spec.UpdateStratgegy.Parition` is not nil, and 
+   `.Spec.UpdateStrategy.Partition.Ordinal` not in the sequence 
+   `(0,.Spec.Replicas)`.
```

Comment:
We already do all the validations I mentioned in Deployments. I am in favor of having them because it will enforce us to fix any culprits but I can see why we wouldn't want them, too.